packages/core/src/linker/template_ref.ts

```diff
@@ -59,11 +69,61 @@ export abstract class TemplateRef<C> {
    * @internal
    * @nocollapse
    */
-  static __NG_ELEMENT_ID__:
-      () => TemplateRef<any>| null = () => SWITCH_TEMPLATE_REF_FACTORY(TemplateRef, ElementRef)
+  static __NG_ELEMENT_ID__: () => TemplateRef<any>| null = SWITCH_TEMPLATE_REF_FACTORY;
 }
 
-export const SWITCH_TEMPLATE_REF_FACTORY__POST_R3__ = render3InjectTemplateRef;
-const SWITCH_TEMPLATE_REF_FACTORY__PRE_R3__ = noop;
-const SWITCH_TEMPLATE_REF_FACTORY: typeof render3InjectTemplateRef =
-    SWITCH_TEMPLATE_REF_FACTORY__PRE_R3__;
+const VE_TemplateRef = TemplateRef;
+
+const R3TemplateRef = class TemplateRef<T> extends VE_TemplateRef<T> {
+  constructor(
+      private _declarationLView: LView, private _declarationTContainer: TContainerNode,
+      public elementRef: ElementRef) {
+    super();
+  }
+
+  createEmbeddedView(context: T): EmbeddedViewRef<T> {
+    const embeddedTView = this._declarationTContainer.tViews as TView;
+    const embeddedLView = createLView(
+        this._declarationLView, embeddedTView, context, LViewFlags.CheckAlways, null,
+        embeddedTView.declTNode, null, null, null, null);
+
+    const declarationLContainer = this._declarationLView[this._declarationTContainer.index];
+    ngDevMode && assertLContainer(declarationLContainer);
+    embeddedLView[DECLARATION_LCONTAINER] = declarationLContainer;
+
+    const declarationViewLQueries = this._declarationLView[QUERIES];
+    if (declarationViewLQueries !== null) {
+      embeddedLView[QUERIES] = declarationViewLQueries.createEmbeddedView(embeddedTView);
+    }
+
+    renderView(embeddedTView, embeddedLView, context);
+
+    return new R3_ViewRef<T>(embeddedLView);
+  }
+};
+
+/**
+ * Creates a TemplateRef given a node.
+ *
+ * @returns The TemplateRef instance to use
+ */
+export function injectTemplateRef<T>(): TemplateRef<T>|null {
+  return createTemplateRef<T>(getCurrentTNode()!, getLView());
+}
+
+/**
+ * Creates a TemplateRef and stores it on the injector.
+ *
+ * @param hostTNode The node on which a TemplateRef is requested
+ * @param hostLView The `LView` to which the node belongs
+ * @returns The TemplateRef instance or null if we can't create a TemplateRef on a given node type
+ */
+export function createTemplateRef<T>(hostTNode: TNode, hostLView: LView): TemplateRef<T>|null {
+  if (hostTNode.type & TNodeType.Container) {
+    ngDevMode && assertDefined(hostTNode.tViews, 'TView must be allocated');
+    return new R3TemplateRef(
+        hostLView, hostTNode as TContainerNode, createElementRef(hostTNode, hostLView));
+  } else {
```

Comment:
this else is unnecessary .
```suggestion
  }
  return null;
```